src/mumble/GlobalShortcut_win.cpp

```diff
@@ -18,80 +26,95 @@
-	/// @param  extended            Indicates whether the button is an extended key in
-	///                             Windows nomenclature. ("[...] such as the right-hand ALT
-	///                             and CTRL keys that appear on an enhanced 101- or 102-key
-	///                             keyboard")
-	/// @param  down                The down/pressed status of the keyboard button.
-	InjectKeyboardMessageEvent(DWORD scancode, DWORD vkcode, bool extended, bool down)
-		: QEvent(static_cast< QEvent::Type >(INJECTKEYBOARDMESSAGE_QEVENT)), m_scancode(scancode), m_vkcode(vkcode),
-		  m_extended(extended), m_down(down) {}
+	/// @param flags      The button flags (i.e. E0/E1 and its state).
+	/// @param scanCode   The Windows scancode of the button.
+	/// @param virtualKey The Windows virtual keycode of the button.
+	InjectKeyboardMessageEvent(const uint16_t flags, const uint16_t scanCode, const uint16_t virtualKey)
+		: QEvent(static_cast< QEvent::Type >(INJECTKEYBOARDMESSAGE_QEVENT)), m_flags(flags), m_scanCode(scanCode),
+		  m_virtualKey(virtualKey) {}
 
 	inline boost::future< bool > shouldSuppress() { return m_suppressionPromise.get_future(); }
 };
 
 /// InjectMouseMessageEvent is an event that can be sent to
 /// the GlobalShortcutWin engine to inject a native Windows mouse
-/// event into GlobalShortcutWin's event stream.
+/// packet into GlobalShortcutWin's event stream.
 class InjectMouseMessageEvent : public QEvent {
-	Q_DISABLE_COPY(InjectMouseMessageEvent);
+	Q_DISABLE_COPY(InjectMouseMessageEvent)
 
 public:
 	boost::promise< bool > m_suppressionPromise;
-	unsigned int m_btn;
-	bool m_down;
+	uint16_t m_flags;
+	uint16_t m_data;
 
 	/// Construct a new InjectMouseMessageEvent.
 	///
-	/// @param  btn                 The DirectInput button index of the mouse event.
-	/// @param  down                The down/pressed status of the mouse button.
-	InjectMouseMessageEvent(unsigned int btn, bool down)
-		: QEvent(static_cast< QEvent::Type >(INJECTMOUSEMESSAGE_QEVENT)), m_btn(btn), m_down(down) {}
+	/// @param flags The button flags (i.e. which button and its state).
+	/// @param data  Specifies the distance the wheel is rotated.
+	InjectMouseMessageEvent(const uint16_t flags, const uint16_t data)
+		: QEvent(static_cast< QEvent::Type >(INJECTMOUSEMESSAGE_QEVENT)), m_flags(flags), m_data(data) {}
 
 	inline boost::future< bool > shouldSuppress() { return m_suppressionPromise.get_future(); }
 };
 
-uint qHash(const GUID &a) {
-	uint val = a.Data1 ^ a.Data2 ^ a.Data3;
-	for (int i = 0; i < 8; i++)
-		val += a.Data4[i];
-	return val;
-}
+/// InjectHIDMessageEvent is an event that can be sent to
+/// the GlobalShortcutWin engine to inject a native Windows HID
+/// packet into GlobalShortcutWin's event stream.
+class InjectHIDMessageEvent : public QEvent {
+	Q_DISABLE_COPY(InjectHIDMessageEvent)
+
+public:
+	boost::promise< bool > m_suppressionPromise;
+	std::unique_ptr< char[] > m_reports;
+	uint32_t m_reportSize;
```

Comment:
Why don't you use a `std::vector` instead?

Reply:
Just because we don't need the extra features provided by `std::vector`.

Reply:
The new code uses `std::vector` and passes the variable with `std::move`.